Various small code fixes / type fixes

bradenmacdonald opened the pull request:
## Description

I created a script to run TypeScript type checking on all the MDX code examples in the docs, and it found these small issues. I will add inline comments to explain where helpful.

### Deploy Preview

https://deploy-preview-3313--paragon-openedx.netlify.app/

## Merge Checklist

n/a - no changes to how any of the Paragon components actually work or look.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@openedx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

bradenmacdonald committed:
docs: fix many small type issues in the docs

bradenmacdonald committed:
fix: missing 'value' prop for <IconButton>

bradenmacdonald committed:
fix: small type error in propTypes utils requiredWhen()

openedx-webhooks commented:
Thanks for the pull request, @bradenmacdonald!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/paragon-working-group`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

bradenmacdonald commented on src/Dropdown/README.md:
```diff
@@ -145,7 +145,7 @@ You can use `Dropdown.Toggle` with [IconButton](/components/iconbutton) componen
 <Dropdown.Deprecated>
   <Dropdown.Deprecated.Button>
-    <Icon className="fa fa-user pr-3" alt="" />
+    <Icon className="fa fa-user pr-3" />
```
[`<Icon>`](https://paragon-openedx.netlify.app/components/icon/#props-api) does not accept an `alt` prop. It accepts an optional `screenReaderText` but it can be omitted for decorative icons, to emulate `alt=""`

PKulkoRaccoonGang replied:
Good catch!

bradenmacdonald commented on src/IconButton/index.tsx:
```diff
@@ -36,6 +36,8 @@ interface Props extends React.HTMLAttributes<HTMLButtonElement> {
   /** size of button to render */
   size?: 'sm' | 'md' | 'inline';
+  /** Used with `IconButtonToggle` */
+  value?: string;
```
As you can see in the docs, [`<IconButtonToggle>` requires its `<IconButton>` children to accept a `value` prop](https://paragon-openedx.netlify.app/components/iconbuttontoggle/), but that wasn't defined in the prop types.

Fixed a type error:

```
Property 'value' does not exist on type 'IntrinsicAttributes & Pick<Pick<PropsWithTooltip, "variant" | "className" | "children" | "slot" | "style" | "title" | "size" | "defaultChecked" | ... 260 more ... | "tooltipContent"> & Pick<...> & Pick<...>, "tooltipContent"> & Partial<...> & Partial<...>'.
```

bradenmacdonald commented on src/Modal/standard-modal.mdx:
```diff
@@ -31,7 +31,7 @@ The standard `ModalDialog` composition. `StandardModal` passes all of its props
           <ActionRow>
             <p className="small">
-              <Hyperlink href="#">Get help</Hyperlink>
+              <Hyperlink destination="#">Get help</Hyperlink>
```
Although this was working, [the `<Hyperlink>` component](https://paragon-openedx.netlify.app/components/hyperlink/) expects `destination` as a prop, not `href`.

bradenmacdonald commented on src/utils/propTypes/utils.js:
```diff
@@ -36,7 +36,7 @@ export const isEveryPropDefined = (props, otherPropNames) => otherPropNames
  * is truthy.
  * @param {func} propType - target PropType
- * @param {string} otherPropName - string name for prop that, if true, marks the
+ * @param {string | string[]} otherPropName - string name for prop that, if true, marks the
```
As you can see in the implementation below as well as how this function is used, this param accepts a string array.

bradenmacdonald commented on src/ListBox/README.md:
```diff
@@ -105,7 +105,7 @@ class ListBoxWrapperForOnSelect extends React.Component {
           ) : (
             <span
-              arialabelledby={`list-box-option-${
+              aria-labelledby={`list-box-option-${
```
`Property 'arialabelledby' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLSpanElement>, HTMLSpanElement>'. Did you mean '"aria-labelledby"'?`

bradenmacdonald commented on src/InputSelect/README.md:
```diff
@@ -90,14 +90,13 @@ notes: |
       };
     }
-    return feedback;
+    return { isValid: true };
```
This change (A) simplifies the validator function, making it a bit more clear, and (B) helps TypeScript avoid a type inference mistake: `Object literal may only specify known properties, and 'validationMessage' does not exist in type '{ isValid: boolean; }'.`

bradenmacdonald commented on src/Form/form-control.mdx:
```diff
@@ -255,10 +257,6 @@ See [react-imask](https://imask.js.org) for documentation on available props.
-  const [value, setValue] = useState('');
-
-  const handleChange = (e) => setValue(e.target.value);
-
```
Fixes: `Block-scoped variable 'value' used before its declaration.`

bradenmacdonald commented on src/DataTable/README.md:
```diff
@@ -253,7 +253,7 @@ To enable proper selection behavior with backend pagination (i.e., when ``isSele
   );
 
-  const ClearAction = ({ as: Component, tableInstance, ...rest }) => (
+  const ClearAction = ({ as: Component, tableInstance }) => (
```
The `rest` parameter was unused.

Fixes: `'rest' is declared but its value is never read.`

bradenmacdonald commented on www/src/components/exampleComponents/HipsterIpsum.tsx:
```diff
@@ -48,18 +48,15 @@ const shortParagraphs = [
+const getHipsterIpsumContent = (allParagraphs: Array<string>, numParagraphs: number) => <>{allParagraphs
   .slice(0, numParagraphs)
-  .map(text => <p key={text}>{text}</p>);
+  .map(text => <p key={text}>{text}</p>)}</>;
```
Fixes:

```
'HipsterIpsum' cannot be used as a JSX component.
  Its return type 'Element[]' is not a valid JSX element.
```

brian-smith-tcril commented on www/src/components/exampleComponents/HipsterIpsum.tsx:
Just one tiny thing here. Can't do an inline suggestion because it's not a line that was changed, but moving away from using `PropTypes` is causing a check failure
> 'PropTypes' is defined but never used

brian-smith-tcril reviewed (commented):
Overall this is looking great. I'm not sure what's going on with the linting errors in `HipsterIpsum` (specifically why they're showing up as annotations on the diff but not failing the lint check)

Other than that this LGTM!

bradenmacdonald commented:
Weird. Yeah, I didn't even notice that check failure because of how it was showing up. Thanks for spotting it. I've rebased this and fixed the lint issues in `HipsterIpsum`.

brian-smith-tcril reviewed (commented):
Thanks for clearing up the lint errors! This LGTM! :shipit:
